doc/dev/backend-plugins.md

```diff
@@ -0,0 +1,271 @@
+  The name of `parentKey` is chosen to make it easier for the plugin to produce good error messages.
+- A keyset `definition` containing the mountpoint definition.
+  To make things easier for the plugin, keys in `definition` are translated into cascading keys relative to `parentKey`.
+  For example, if the key `system:/elektra/mountpoints/system:\/hosts/path` is set in the KDB, then `definition` will contain a key `/path`.
+
+TODO: how does the backend plugin get access to the other plugins? Maybe, just add a plugins/# array into `ks` containing keys with `Plugin *` values?
+
+The backend plugin then:
+
+- **MUST** parse the mountpoint definition and store the necessary information for later phases internally.
+- **SHOULD** validate that the mountpoint definition _can be_ valid.
+- **SHOULD NOT** do other validation.
+  For example a file-based backend, _should not_ check whether the file(s) or path(s) referenced in the mountpoint definition exist.
+  Such a check should be done in the `resolver` phase.
+- **MAY** decided that the mountpoint should be read-only.
+  If so, this must be indicated to `libelektra-kdb` via the return value.
+
+This phase exists purely for the backend plugin to initialize and configure itself for the mountpoint.
+
+> **Note**: This phase is only executed _once per instance of `KDB`_.
+> Only the first `kdbGet()` call will result in `libelektra-kdb` executing this phase, all future calls to `kdbGet()` (and `kdbSet()`) start with the `resolver` phase.
+> The backend plugin must store the information contained in the mountpoint definition internally to accommodate this.
+
+#### Resolver Phase
+
+During the `resolver` phase the backend plugin is called with:
+
+- A key `parentKey` whose name is the parent key of the mountpoint and whose value is an empty string.
+  The key name of this key is read-only.
+- An empty keyset `ks`.
+
+The backend plugin then:
+
+- **MUST** set the value of the `parentKey` to a value identifying the storage unit that contains the data of the mountpoint.
+  For file-based backend plugins, this means setting the value of `parentKey` to an absolute filename.
+- **MAY** set metadata on `parentKey`, if encoding the information required for the following phases is too hard to encode in a single string value.
+
+> **Note**: The backend plugin may also modify the keyset `ks`, but `libelektra-kdb` will discard this keyset after this phase, so these modifications won't have any effects.
+
+During a `set` operation the backend plugin must also ensure that errors in the storage phases can be safely rolled back and that the `set` operation does not affect concurrent operations.
+For file-based backends, this means creating a temporary storage file and returning its absolute filename instead of the name of the actual storage file.
+In other words, in a `set` operation the `resolver` phase is also about preparing a transaction in addition to resolving the storage unit.
+
+#### Cache-Check Phase
+
```

Comment:
As noted below, this comparison code is not the same for all resolvers. This is only one implementation and I'm pretty sure this should stay in each resolver, because different resolvers might have different notions of file change.

Reply:
> As noted below, this comparison code is not the same for all resolvers. This is only one implementation and I'm pretty sure this should stay in each resolver, because different resolvers might have different notions of file change.

The idea was that the `backend` plugin would only be for file-based backends. For e.g. a git-based backend or a database backend you would use a different plugin. These other plugins could do different comparisons. There may be a different resolver plugin involved, but if the resolver code is simple enough then it might be hardcoded without a different plugin.

That said, I will probably keep this code in `resolver` but export it as a separate function that is called by `backend`. That makes it clearer what is happening and would still allow a different file-based resolver (e.g. `wresolver`) to use a different time check.

Reply:
I agree that we should focus on file-based backends for this PR and 1.0. Later we might create new backends without resolvers, or different non-file-based resolving mechanisms.